### ISSUE TITLE
```query``` endpoint returning "empty" values

### ISSUE THREAD

**Issue by frigaardj** [OpenTSDB/opentsdb#540]
Running OpenTSDB 2.1, the query endpoint returns a mix of valid & empty metrics when ```groupBy = true```. I have tried running ```uid metasync``` and changing the query parameters, but the values still come up:
```json
  [{
    "metric": "themetricname",
    "tags": {
      "id": "dedupe",
      "host": "host1"
    },
    "aggregateTags": [],
    "dps": {}
  },
  {
    "metric": "themetricname",
    "tags": {
      "id": "Document_Download",
      "host": "aflow"
    },
    "aggregateTags": [],
    "dps": {
      "1438326573": 1438326000000
    }
  },
  {
    "metric": "",
    "tags": {},
    "aggregateTags": [],
    "dps": {}
  },
  ...
  ]
```

This was achieved with the following query:

```
http://_________/api/query/?groupBy=true&m=avg:themetricname{id=*,host=*}&start=1438323044&end=1438326644
```

This query returns about a thousand unique id/host combinations, and about 100 "empty" metrics with no name, tags, dps. I don't think there is anything abnormal about the underlying data, and irrespective this endpoint should not be returning values of ```"metric": ""``` since we queried a specific metric. Can sidestep this by post-filtering but would be nice to resolve the bug.

**Comment by bhinav01** [comment]
Facing the same issue with OpenTSDB 2.0. We don't use groupBy. Here's the result from one of the queries:
[
...
{
	"metric": "",
	"tags": {},
	"aggregateTags": [],
	"dps": {}
},
{
	"metric": "production",
	"tags": {
		"orgId": "bla1",
		"podId": "bla2",
		"meta": "bla3"
	},
	"aggregateTags": [],
	"dps": {
		"1439223600": 203.1597442626953,
		"1439224200": 145.2007598876953,
		"1439224800": 112.61372375488281,
		"1439225400": 133.38345336914062,
		"1439226000": 133.10476684570312,
		"1439226600": 112.19779968261719
	}
}
...
]

The query is of the form:
http://myhost:myport/api/query?start=4h-ago&end=3h-ago&m=avg:production{metric=apt,orgId=*,podId=*}

This query returns a mix of valid metrics and empty metrics with no name, tags, dps. Would be nice to see this resolved.

**Comment by manolama** [comment]
Yeah, this is a bit of ugliness that should be fixed. What happens is that TSD scans for data a bit before and after your query time span to account for downsampling and rate calculations. When we start writing the output, we "skip" any data points outside the range so if there was a value before the query start time, we'll see there was a series there, and will have written the start of the output but won't finish it. We'll fix it.
